mxfusion/inference/score_function.py

```diff
@@ -0,0 +1,187 @@
+
+class ScoreFunctionRBInference(ScoreFunctionInference):
+    """
+    Implemented following the [Black Box Variational Inference](https://arxiv.org/abs/1401.0118) paper.
+
+    The addition of Rao-Blackwellization and Control Variates (RBCV) requires that the posterior passed in be of meanfield form (i.e. all posterior variables independent.)
+
+    Terminology:
+      Lambda - Posterior parameters
+      Theta - Model parameters
+
+    :param num_samples: the number of samples used in estimating the variational lower bound
+    :type num_samples: int
+    :param model: the definition of the probabilistic model
+    :type model: Model
+    :param posterior: the definition of the variational posterior of the probabilistic model
+    :param posterior: Posterior
+    :param observed: A list of observed variables
+    :type observed: [Variable]
+    """
+
+    def __init__(self, num_samples, model, posterior, observed):
+        super(ScoreFunctionRBInference, self).__init__(
+            model=model, observed=observed, posterior=posterior, num_samples=num_samples)
+
+    def compute(self, F, data, parameters, constants):
+        """
+        The method for the computation of the inference algorithm
+
+        :param F: the execution context (mxnet.ndarray or mxnet.symbol)
+        :type F: Python module
+        :param data: the data variables for inference
+        :type data: {Variable: mxnet.ndarray.ndarray.NDArray or
+            mxnet.symbol.symbol.Symbol}
+        :param parameters: the parameters for inference
+        :type parameters: {Variable: mxnet.ndarray.ndarray.NDArray or
+            mxnet.symbol.symbol.Symbol}
+        :param constants: the constants for inference
+        :type parameters: {Variable: mxnet.ndarray.ndarray.NDArray or
+            mxnet.symbol.symbol.Symbol}
+        :returns: the outcome of the inference algorithm
+        :rtype: mxnet.ndarray.ndarray.NDArray or mxnet.symbol.symbol.Symbol
+        """
+        knowns = data.copy()
+        knowns.update(parameters)
+        knowns.update(constants)
+
+        # Sample q(z | lambda)
+        samples = self.posterior.draw_samples(
+            F=F, conditionals=knowns, num_samples=self.num_samples,
+            constants=constants)
+        knowns.update(samples)
+
+        q_z_lambda = self.posterior.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        p_x_z = self.model.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        gradient_theta = F.mean(p_x_z - F.stop_gradient(q_z_lambda), axis=0)
+        posterior_rvs = [v for v in self.posterior.variables.values() if v.type is VariableType.RANDVAR]
+        f_list = []
+        grad_list = []
+        for i, v in enumerate(posterior_rvs):
+            model_v = self.model[v]
+
+            q_i_varset = self._extract_descendant_blanket_params(self.posterior, v)
+            q_i_params = {key:val for key,val in knowns.items() if key in q_i_varset}
```

Comment:
Why not do the following?
```python
q_i_params = {key: knowns[key] for key in q_i_varset}
```

Reply:
This isn't the same, since not all variables in q_i_varset are in knowns (i.e. the unknown ones.)